File: analysis/pearlfortune/ghidra_scripts/AnalyzePCIe.java
//Ghidra script to find PCIe Gen3 references
//@category Analysis
//@keybinding
//@menupath
//@toolbar

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.address.*;
import ghidra.program.model.mem.*;

public class AnalyzePCIe extends GhidraScript {
    
    @Override
    public void run() throws Exception {
        println("============================================================");
        println("PCIe Gen3 Analysis for cmpunlocker-rs");
        println("============================================================");
        
        Memory memory = currentProgram.getMemory();
        
        // Search strings
        String[] patterns = {
            "EnablePCIeGen3",
            "NVreg_EnablePCIeGen3",
            "RMPcieLinkSpeed",
            "Gen3",
            "kbifGetPciLinkMaxSpeedByPciGenInfo",
            "calculatePCIELinkRateMBps",
            "XVE_LINK_CONTROL",
            "pcie_cap",
            "nvidia-modprobe",
            "modprobe",
        };
        
        println("\n[1] STRING SEARCH:");
        for (String pattern : patterns) {
            if (monitor.isCancelled()) return;
            
            byte[] bytes = pattern.getBytes();
            Address addr = memory.findBytes(currentProgram.getMinAddress(), bytes, null, true, monitor);
            
            int count = 0;
            while (addr != null && count < 5) {
                println("  '" + pattern + "' @ " + addr);
                
                // Find xrefs
                Reference[] refs = getReferencesTo(addr);
                for (int i = 0; i < refs.length && i < 3; i++) {
                    Address from = refs[i].getFromAddress();
                    Function func = getFunctionContaining(from);
                    String fname = (func != null) ? func.getName() : "?";
                    println("    <- " + from + " in " + fname);
                }
                
                count++;
                addr = memory.findBytes(addr.add(1), bytes, null, true, monitor);
            }
        }
        
        // Search register addresses
        println("\n[2] REGISTER ADDRESSES (LE):");
        long[] regs = {
            0x00823810L,  // PCIE_FUSE
            0x0082057cL,  // OPT_GEN23
            0x00088088L,  // XVE LnkCap
            0x0008c000L,  // LINK_CONTROL
            0x0008c040L,  // LINK_SPEED
            0x0082381cL,  // SS0
            0x00823820L,  // SS1
        };
        String[] regNames = {"PCIE_FUSE", "OPT_GEN23", "XVE_LnkCap", "LINK_CONTROL", "LINK_SPEED", "SS0", "SS1"};
        
        for (int r = 0; r < regs.length; r++) {
            if (monitor.isCancelled()) return;
            
            long reg = regs[r];
            byte[] bytes = new byte[4];
            bytes[0] = (byte)(reg & 0xFF);
            bytes[1] = (byte)((reg >> 8) & 0xFF);
            bytes[2] = (byte)((reg >> 16) & 0xFF);
            bytes[3] = (byte)((reg >> 24) & 0xFF);
            
            Address addr = memory.findBytes(currentProgram.getMinAddress(), bytes, null, true, monitor);
            int count = 0;
            while (addr != null && count < 3) {
                println("  " + regNames[r] + " (0x" + Long.toHexString(reg) + ") @ " + addr);
                
                Reference[] refs = getReferencesTo(addr);
                for (int i = 0; i < refs.length && i < 2; i++) {
                    Address from = refs[i].getFromAddress();
                    Function func = getFunctionContaining(from);
                    String fname = (func != null) ? func.getName() : "?";
                    println("    <- " + from + " in " + fname);
                }
                
                count++;
                addr = memory.findBytes(addr.add(1), bytes, null, true, monitor);
            }
        }
        
        // List PCIe-related functions
        println("\n[3] PCIE/GEN3-RELATED FUNCTIONS:");
        FunctionManager fm = currentProgram.getFunctionManager();
        FunctionIterator funcs = fm.getFunctions(true);
        int funcCount = 0;
        while (funcs.hasNext() && funcCount < 50) {
            Function func = funcs.next();
            String name = func.getName().toLowerCase();
            if (name.contains("pcie") || name.contains("gen3") || name.contains("link") || 
                name.contains("xve") || name.contains("bifurcation") || name.contains("nvidia")) {
                println("  " + func.getName() + " @ " + func.getEntryPoint());
                funcCount++;
            }
        }
        
        // Decompile main or entry
        println("\n[4] LOOKING FOR MAIN/ENTRY:");
        Function main = null;
        for (Function f : fm.getFunctions(true)) {
            if (f.getName().equals("main") || f.getName().equals("_start") || 
                f.getName().contains("::main")) {
                main = f;
                println("  Found: " + f.getName() + " @ " + f.getEntryPoint());
            }
        }
        
        println("\n============================================================");
        println("Analysis complete. Use Ghidra GUI for detailed decompilation.");
        println("============================================================");
    }
}

File: analysis/pearlfortune/ghidra_scripts/analyze_pcie.py
# Ghidra Python script to find PCIe Gen3 references in cmpunlocker-rs
# @category Analysis

from ghidra.program.model.listing import CodeUnit
from ghidra.program.model.symbol import RefType
import re

def search_strings(program, patterns):
    """Search for string patterns in memory"""
    memory = program.getMemory()
    results = []
    
    for pattern in patterns:
        addr = memory.findBytes(program.getMinAddress(), 
                               pattern.encode(), None, True, monitor)
        while addr and not monitor.isCancelled():
            results.append((pattern, addr))
            addr = memory.findBytes(addr.add(1), pattern.encode(), None, True, monitor)
    
    return results

def get_xrefs(program, addr):
    """Get all references to an address"""
    refs = []
    refMgr = program.getReferenceManager()
    for ref in refMgr.getReferencesTo(addr):
        refs.append(ref.getFromAddress())
    return refs

def main():
    print("=" * 60)
    print("PCIe Gen3 Analysis for cmpunlocker-rs")
    print("=" * 60)
    
    # Search patterns
    patterns = [
        "EnablePCIeGen3",
        "NVreg_EnablePCIeGen3",
        "RMPcieLinkSpeed",
        "kbifGetPciLinkMaxSpeedByPciGenInfo",
        "calculatePCIELinkRateMBps",
        "Gen3",
        "pcie",
        "PCIe",
        "LinkSpeed",
        "0x823810",  # PCIE_FUSE
        "0x82057c",  # OPT_GEN23
    ]
    
    print("\n[1] String Search Results:")
    found = search_strings(currentProgram, patterns)
    for pattern, addr in found:
        print(f"  '{pattern}' @ {addr}")
        xrefs = get_xrefs(currentProgram, addr)
        for xref in xrefs[:5]:  # Limit xrefs
            func = getFunctionContaining(xref)
            fname = func.getName() if func else "unknown"
            print(f"    <- {xref} in {fname}")
    
    # Search for register constants
    print("\n[2] Register Constants (little-endian):")
    registers = [
        (0x00823810, "PCIE_FUSE"),
        (0x0082057c, "OPT_GEN23"),
        (0x00088088, "XVE_LnkCap"),
        (0x0008c000, "LINK_CONTROL"),
        (0x0008c040, "LINK_SPEED"),
    ]
    
    memory = currentProgram.getMemory()
    for reg, name in registers:
        # Little-endian 4-byte search
        bytes_le = bytearray([
            reg & 0xFF,
            (reg >> 8) & 0xFF,
            (reg >> 16) & 0xFF,
            (reg >> 24) & 0xFF
        ])
        addr = memory.findBytes(currentProgram.getMinAddress(), bytes(bytes_le), None, True, monitor)
        count = 0
        while addr and count < 5:
            print(f"  {name} (0x{reg:08x}) @ {addr}")
            xrefs = get_xrefs(currentProgram, addr)
            for xref in xrefs[:3]:
                func = getFunctionContaining(xref)
                fname = func.getName() if func else "unknown"
                print(f"    <- {xref} in {fname}")
            count += 1
            addr = memory.findBytes(addr.add(1), bytes(bytes_le), None, True, monitor)
    
    # List functions with "pcie" or "gen3" in name
    print("\n[3] Functions with PCIe-related names:")
    fm = currentProgram.getFunctionManager()
    for func in fm.getFunctions(True):
        name = func.getName().lower()
        if "pcie" in name or "gen3" in name or "link" in name:
            print(f"  {func.getName()} @ {func.getEntryPoint()}")
    
    print("\n" + "=" * 60)
    print("Analysis complete")
    print("=" * 60)

main()
